File: ui_selenium_bdd/agents/observability-agent.py
# ============================================================
# Observability Agent — Traces · Circuit Breaker · Métriques
# ============================================================
# 100% déterministe — zéro appel LLM
#
# Commandes :
#   python agents/observability-agent.py traces         → appels LLM récents
#   python agents/observability-agent.py metrics        → métriques par agent
#   python agents/observability-agent.py cost           → estimation coût tokens
#   python agents/observability-agent.py circuit        → état circuit breaker
#   python agents/observability-agent.py memory         → épisodes mémoire
#   python agents/observability-agent.py prompts list   → prompts versionnés
#   python agents/observability-agent.py prompts rollback <name>
#   python agents/observability-agent.py dashboard      → rapport HTML
# ============================================================

import sys, os, json, glob, time
sys.stdout.reconfigure(encoding="utf-8", errors="replace")
sys.path.insert(0, os.path.dirname(__file__))

from prompt_store import PromptStore

FRAMEWORK = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
LOGS_DIR  = os.path.join(FRAMEWORK, "logs")
DOCS_DIR  = os.path.join(FRAMEWORK, "docs")
TRACES    = os.path.join(LOGS_DIR, "traces.jsonl")
CB_STATE  = os.path.join(LOGS_DIR, "circuit_breaker_state.json")
CACHE     = os.path.join(LOGS_DIR, "llm_cache.json")
MEMORY    = os.path.join(FRAMEWORK, "memory", "episodes.jsonl")

R = "\033[31m"; G = "\033[32m"; Y = "\033[33m"; C = "\033[36m"
W = "\033[1m";  E = "\033[0m"

COST_INPUT  = 0.59 / 1_000_000
COST_OUTPUT = 0.79 / 1_000_000


def load_traces(n: int = 50) -> list:
    if not os.path.exists(TRACES):
        return []
    lines = []
    with open(TRACES, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line:
                try:
                    lines.append(json.loads(line))
                except Exception:
                    pass
    return lines[-n:]


def cmd_traces(n: int = 20):
    print(f"\n{W}OBSERVABILITY — Dernières {n} traces LLM{E}\n")
    traces = load_traces(n)
    if not traces:
        print(f"  {Y}Aucune trace. Lance un agent pour générer des traces.{E}")
        return
    for t in traces[-n:]:
        ok    = t.get("success", True)
        color = G if ok else R
        fn    = t.get("fn", "?")[:25]
        dur   = t.get("durationMs", 0)
        conf  = t.get("confidence")
        conf_str = f" conf={conf:.2f}" if conf is not None else ""
        print(f"  {color}{'✓' if ok else '✗'}{E} {fn:<26} {dur:>5}ms{conf_str}")


def cmd_metrics():
    print(f"\n{W}OBSERVABILITY — Métriques par agent{E}\n")
    traces = load_traces(500)
    if not traces:
        print(f"  {Y}Aucune trace disponible.{E}")
        return
    by_agent = {}
    for t in traces:
        agent = t.get("fn", "unknown").split("_")[0]
        by_agent.setdefault(agent, {"calls":0,"errors":0,"total_ms":0,"confidences":[]})
        by_agent[agent]["calls"] += 1
        if not t.get("success", True):
            by_agent[agent]["errors"] += 1
        by_agent[agent]["total_ms"] += t.get("durationMs", 0)
        if t.get("confidence") is not None:
            by_agent[agent]["confidences"].append(t["confidence"])

    print(f"  {'Agent':<20} {'Appels':>7} {'Erreurs':>8} {'Avg ms':>8} {'Avg conf':>10}")
    print(f"  {'-'*60}")
    for agent, m in sorted(by_agent.items()):
        avg_ms   = int(m["total_ms"] / m["calls"]) if m["calls"] else 0
        avg_conf = round(sum(m["confidences"]) / len(m["confidences"]), 2) if m["confidences"] else "—"
        err_color = R if m["errors"] > 0 else G
        print(f"  {agent:<20} {m['calls']:>7} {err_color}{m['errors']:>8}{E} {avg_ms:>8} {str(avg_conf):>10}")


def cmd_cost():
    print(f"\n{W}OBSERVABILITY — Estimation coût LLM{E}")
    print(f"  Tarif Groq : ${COST_INPUT*1e6:.2f}/M tokens input · ${COST_OUTPUT*1e6:.2f}/M tokens output\n")
    traces = load_traces(1000)
    if not traces:
        print(f"  {Y}Aucune trace.{E}")
        return
    total_in  = sum(t.get("promptLen", 0) // 4 for t in traces)
    total_out = sum(t.get("responseLen", 0) // 4 for t in traces)
    cost = total_in * COST_INPUT + total_out * COST_OUTPUT
    print(f"  Appels totaux : {len(traces)}")
    print(f"  Tokens input  : ~{total_in:,}")
    print(f"  Tokens output : ~{total_out:,}")
    print(f"  Coût estimé   : {G}${cost:.4f}{E}")


def cmd_circuit():
    print(f"\n{W}OBSERVABILITY — Circuit Breaker{E}\n")
    if not os.path.exists(CB_STATE):
        print(f"  {Y}Pas d'état CB (aucun appel LLM encore).{E}")
        return
    with open(CB_STATE, encoding="utf-8") as f:
        cb = json.load(f)
    state = cb.get("state", "CLOSED")
    color = G if state == "CLOSED" else (Y if state == "HALF_OPEN" else R)
    print(f"  État : {color}{W}{state}{E}")
    print(f"  Échecs consécutifs : {cb.get('failure_count', 0)}")
    if cb.get("opened_at"):
        print(f"  Ouvert depuis : {cb['opened_at']}")
    if os.path.exists(CACHE):
        with open(CACHE, encoding="utf-8") as f:
            cache = json.load(f)
        print(f"  Cache LLM : {len(cache)} entrées")


def cmd_memory(n: int = 10):
    print(f"\n{W}OBSERVABILITY — Mémoire épisodique (derniers {n}){E}\n")
    if not os.path.exists(MEMORY):
        print(f"  {Y}Aucun épisode en mémoire.{E}")
        return
    episodes = []
    with open(MEMORY, encoding="utf-8") as f:
        for line in f:
            try:
                episodes.append(json.loads(line.strip()))
            except Exception:
                pass
    for ep in episodes[-n:]:
        ts    = ep.get("timestamp", "?")[:16]
        agent = ep.get("agent", "?")
        summ  = ep.get("summary", "?")[:60]
        print(f"  {C}{ts}{E} [{agent}] {summ}")


def cmd_prompts(sub: str = "list", name: str = None):
    _ps = PromptStore()
    if sub == "list":
        print(f"\n{W}OBSERVABILITY — Prompts versionnés{E}\n")
        prompts = _ps.list_all()
        if not prompts:
            print(f"  {Y}Aucun prompt dans prompts/.{E}")
            return
        print(f"  {'Nom':<22} {'Version':>9} {'Appels':>8} {'Avg conf':>10} {'Desc'}")
        print(f"  {'-'*75}")
        for p in prompts:
            m = p.get("metrics", {})
            calls = m.get("calls", 0)
            conf  = f"{m['avg_confidence']:.2f}" if m.get("avg_confidence") else "—"
            print(f"  {p['name']:<22} {p['current_version']:>9} {calls:>8} {conf:>10}  {p.get('description','')[:35]}")
    elif sub == "rollback" and name:
        try:
            prev = _ps.rollback(name)
            print(f"  {G}Rollback '{name}' → v{prev}{E}")
        except Exception as ex:
            print(f"  {R}Erreur : {ex}{E}")
    elif sub == "versions" and name:
        versions = _ps.list_versions(name)
        for v in versions:
            marker = " ← current" if v.get("is_current") else ""
            print(f"  v{v['version']}  {v.get('created_at','?')[:10]}  {v.get('note','')}{marker}")


def cmd_dashboard():
    print(f"\n{W}OBSERVABILITY — Rapport HTML{E}")
    traces  = load_traces(1000)
    n_calls = len(traces)
    errors  = sum(1 for t in traces if not t.get("success", True))
    total_ms = sum(t.get("durationMs", 0) for t in traces)
    avg_ms  = int(total_ms / n_calls) if n_calls else 0

    html = f"""<!DOCTYPE html><html><head><meta charset="utf-8">
<title>Observability — ui_selenium_bdd</title>
<style>body{{background:#0d1117;color:#c9d1d9;font-family:'Segoe UI',sans-serif;padding:20px}}
.grid{{display:grid;grid-template-columns:repeat(4,1fr);gap:16px;margin:20px 0}}
.card{{background:#161b22;border:1px solid #30363d;border-radius:8px;padding:20px;text-align:center}}
.metric{{font-size:2em;font-weight:bold;margin:8px 0}}h1{{color:#58a6ff}}</style></head>
<body><h1>📡 Observability Dashboard — ui_selenium_bdd</h1>
<p style="color:#8b949e">Généré le {time.strftime('%Y-%m-%d %H:%M')}</p>
<div class="grid">
  <div class="card"><div class="metric" style="color:#58a6ff">{n_calls}</div><div>Appels LLM totaux</div></div>
  <div class="card"><div class="metric" style="color:#e74c3c">{errors}</div><div>Erreurs</div></div>
  <div class="card"><div class="metric" style="color:#2ecc71">{avg_ms}ms</div><div>Durée moyenne</div></div>
  <div class="card"><div class="metric" style="color:#f39c12">${(n_calls * avg_ms / 1000 * 0.001):.4f}</div><div>Coût estimé</div></div>
</div></body></html>"""

    os.makedirs(DOCS_DIR, exist_ok=True)
    with open(os.path.join(DOCS_DIR, "observability-report.html"), "w", encoding="utf-8") as f:
        f.write(html)
    print(f"  {G}Rapport : docs/observability-report.html{E}")


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="Observability Agent")
    parser.add_argument("command", choices=["traces","metrics","cost","circuit","memory","prompts","dashboard"])
    parser.add_argument("sub", nargs="?", default="list")
    parser.add_argument("name", nargs="?", default=None)
    parser.add_argument("-n", type=int, default=20)
    args = parser.parse_args()

    if args.command == "traces":    cmd_traces(args.n)
    elif args.command == "metrics": cmd_metrics()
    elif args.command == "cost":    cmd_cost()
    elif args.command == "circuit": cmd_circuit()
    elif args.command == "memory":  cmd_memory(args.n)
    elif args.command == "prompts": cmd_prompts(args.sub, args.name)
    elif args.command == "dashboard": cmd_dashboard()

File: ui_cypress_bdd/scripts/agents/observability-agent.js
'use strict';
// ============================================================
// Observability Agent — Monitoring LLM + Résilience + Prompt Versions
// ============================================================
// Usage:
//   node scripts/agents/observability-agent.js metrics       Métriques appels LLM
//   node scripts/agents/observability-agent.js anomalies     Détecte appels lents/erreurs
//   node scripts/agents/observability-agent.js cost          Estimation coût tokens
//   node scripts/agents/observability-agent.js cb status     État du Circuit Breaker
//   node scripts/agents/observability-agent.js cb reset      Réinitialise le CB
//   node scripts/agents/observability-agent.js cb cache      Affiche stats cache LLM
//   node scripts/agents/observability-agent.js prompts list  Liste les prompts versionnés
//   node scripts/agents/observability-agent.js prompts history <name>  Historique d'un prompt
//   node scripts/agents/observability-agent.js prompts rollback <name> Rollback version
//   node scripts/agents/observability-agent.js report        Génère docs/observability-report.html
//   node scripts/agents/observability-agent.js clear         Vide les traces
//
// Coût LLM : zéro (100% déterministe — lecture de fichiers et calculs)
// ============================================================
require('dotenv').config();

const fs     = require('fs');
const path   = require('path');
const tracer = require('./shared/tracer');
const cb     = require('./shared/circuit-breaker');
const ps     = require('./shared/prompt-store');

const FRAMEWORK = path.join(__dirname, '..', '..');
const DOCS_DIR  = path.join(FRAMEWORK, 'docs');
fs.mkdirSync(DOCS_DIR, { recursive: true });

const PRICING = { input: 0.59, output: 0.79 };

const SLOW_MS      = 8000;
const VERY_SLOW_MS = 15000;
const LOW_CONF     = 0.65;
const ERR_BURST    = 3;

const G = '\x1b[32m', R = '\x1b[31m', Y = '\x1b[33m', C = '\x1b[36m', B = '\x1b[1m', E = '\x1b[0m';

// ── METRICS ───────────────────────────────────────────────────────────────────
function cmdMetrics() {
  console.log(`\n${B}=== OBSERVABILITY — METRICS ===${E}`);
  const traces = tracer.loadTraces();
  if (!traces.length) { console.log(`  ${Y}⚠  Aucune trace dans logs/traces.jsonl${E}`); return; }

  const byAgent = {};
  for (const t of traces) {
    const key = `${t.agent}/${t.fn}`;
    if (!byAgent[key]) byAgent[key] = { calls: 0, totalMs: 0, errors: 0, durations: [] };
    const s = byAgent[key];
    s.calls++; s.totalMs += t.duration_ms || 0;
    if (!t.success) s.errors++;
    s.durations.push(t.duration_ms || 0);
  }

  console.log(`\n  ${C}${traces.length} traces — ${Object.keys(byAgent).length} fonctions uniques${E}\n`);
  console.log(`  ${'Agent/Fonction'.padEnd(40)} ${'Calls'.padEnd(7)} ${'Moy ms'.padEnd(9)} ${'P95 ms'.padEnd(9)} Erreurs`);
  console.log(`  ${'─'.repeat(78)}`);

  for (const [key, s] of Object.entries(byAgent).sort((a,b) => b[1].calls - a[1].calls)) {
    const avgMs = Math.round(s.totalMs / s.calls);
    const sorted = [...s.durations].sort((a,b) => a-b);
    const p95    = sorted[Math.floor(sorted.length * 0.95)] || 0;
    const errPct = Math.round(s.errors / s.calls * 100);
    const errStr = s.errors ? R + `${s.errors} (${errPct}%)` + E : G + '0' + E;
    const msColor = avgMs > SLOW_MS ? R : avgMs > 3000 ? Y : G;
    console.log(`  ${key.padEnd(40)} ${String(s.calls).padEnd(7)} ${msColor}${String(avgMs).padEnd(9)}${E} ${String(Math.round(p95)).padEnd(9)} ${errStr}`);
  }
}

// ── ANOMALIES ─────────────────────────────────────────────────────────────────
function cmdAnomalies() {
  console.log(`\n${B}=== OBSERVABILITY — ANOMALIES ===${E}`);
  const traces  = tracer.loadTraces();
  if (!traces.length) { console.log(`  ${Y}⚠  Aucune trace${E}`); return; }

  const anomalies = [];
  const recentErrors = traces.slice(-20).filter(t => !t.success);

  for (const t of traces) {
    if (t.duration_ms >= VERY_SLOW_MS) anomalies.push({ type: 'very_slow', trace: t, msg: `${Math.round(t.duration_ms/1000)}s — ${t.agent}/${t.fn}` });
    else if (t.duration_ms >= SLOW_MS) anomalies.push({ type: 'slow',      trace: t, msg: `${Math.round(t.duration_ms/1000)}s — ${t.agent}/${t.fn}` });
    if (t.confidence != null && t.confidence < LOW_CONF) anomalies.push({ type: 'low_conf', trace: t, msg: `conf=${t.confidence} — ${t.agent}/${t.fn}` });
  }
  if (recentErrors.length >= ERR_BURST) anomalies.push({ type: 'error_burst', msg: `${recentErrors.length} erreurs sur les 20 derniers appels` });

  if (!anomalies.length) { console.log(`  ${G}✓ Aucune anomalie détectée${E}`); return; }
  console.log(`  ${R}${anomalies.length} anomalie(s) détectée(s)${E}\n`);

  for (const a of anomalies) {
    const icon = a.type === 'very_slow' ? R+'🔴'+E : a.type === 'slow' ? Y+'🟡'+E : a.type === 'low_conf' ? Y+'⚡'+E : R+'💥'+E;
    console.log(`  ${icon}  [${a.type.toUpperCase()}]  ${a.msg}`);
  }
}

// ── COST ──────────────────────────────────────────────────────────────────────
function cmdCost() {
  console.log(`\n${B}=== OBSERVABILITY — COÛT TOKENS ===${E}`);
  const traces = tracer.loadTraces();
  if (!traces.length) { console.log(`  ${Y}⚠  Aucune trace${E}`); return; }

  let totalInput = 0, totalOutput = 0;
  const byAgent  = {};

  for (const t of traces) {
    const inp = t.prompt_len  / 4;
    const out = t.response_len / 4;
    totalInput  += inp; totalOutput += out;
    if (!byAgent[t.agent]) byAgent[t.agent] = { input: 0, output: 0, calls: 0 };
    byAgent[t.agent].input  += inp;
    byAgent[t.agent].output += out;
    byAgent[t.agent].calls++;
  }

  const costInput  = totalInput  / 1_000_000 * PRICING.input;
  const costOutput = totalOutput / 1_000_000 * PRICING.output;
  const total      = costInput + costOutput;

  console.log(`\n  ${B}Total estimé (Groq llama-3.3-70b) :${E} $${total.toFixed(4)}`);
  console.log(`  Input : ~${Math.round(totalInput).toLocaleString()} tokens  ($${costInput.toFixed(4)})`);
  console.log(`  Output: ~${Math.round(totalOutput).toLocaleString()} tokens ($${costOutput.toFixed(4)})\n`);

  console.log(`  ${'Agent'.padEnd(30)} ${'Calls'.padEnd(7)} ${'Tokens'.padEnd(12)} Coût estimé`);
  console.log(`  ${'─'.repeat(65)}`);
  for (const [agent, s] of Object.entries(byAgent).sort((a,b) => (b[1].input+b[1].output)-(a[1].input+a[1].output))) {
    const tokens = Math.round(s.input + s.output);
    const cost   = (s.input / 1_000_000 * PRICING.input + s.output / 1_000_000 * PRICING.output);
    console.log(`  ${agent.padEnd(30)} ${String(s.calls).padEnd(7)} ${String(tokens).padEnd(12)} $${cost.toFixed(4)}`);
  }
  console.log(`\n  ${C}💡 Ollama local → $0.00 (toutes les traces)${E}`);
}

// ── CIRCUIT BREAKER ───────────────────────────────────────────────────────────
function cmdCb(action) {
  if (action === 'reset') {
    cb.reset(); console.log(`  ${G}✓ Circuit Breaker réinitialisé → CLOSED${E}`);
  } else if (action === 'cache') {
    const status = cb.getStatus();
    console.log(`\n${B}=== CB — CACHE ===${E}`);
    console.log(`  Entrées en cache : ${status.cacheSize} / ${status.config.cacheMaxEntries}`);
    console.log(`  TTL              : ${status.config.cacheTtlSeconds}s`);
    console.log(`  Fichier          : ${cb.CACHE_FILE}`);
  } else {
    const status = cb.getStatus();
    const stateIcon = status.state === 'CLOSED' ? G+'CLOSED'+E : status.state === 'OPEN' ? R+'OPEN'+E : Y+'HALF_OPEN'+E;
    console.log(`\n${B}=== OBSERVABILITY — CIRCUIT BREAKER ===${E}`);
    console.log(`  État     : ${stateIcon}`);
    console.log(`  Échecs   : ${status.failures} / ${status.config.failureThreshold}`);
    console.log(`  Succès   : ${status.successes} / ${status.config.successThreshold}`);
    console.log(`  Cooldown : ${status.config.cooldownSeconds}s`);
    console.log(`  Cache    : ${status.cacheSize} entrées`);
    if (status.openedAt) console.log(`  Ouvert le: ${new Date(status.openedAt * 1000).toLocaleString('fr-FR')}`);
  }
}

// ── PROMPTS ───────────────────────────────────────────────────────────────────
function cmdPrompts(action, name) {
  if (action === 'list') {
    const prompts = ps.listAll();
    console.log(`\n${B}=== PROMPTS — LISTE (${prompts.length}) ===${E}\n`);
    if (!prompts.length) { console.log(`  ${Y}⚠  Aucun prompt stocké${E}`); return; }
    prompts.forEach(p => console.log(`  ${p.name.padEnd(30)} v${p.current_version.padEnd(8)} ×${p.metrics?.calls||0} appels  ${p.description?.slice(0,35)||''}`));
  } else if (action === 'history') {
    if (!name) { console.log(`  Usage: observability prompts history <name>`); return; }
    const versions = ps.listVersions(name);
    console.log(`\n${B}=== PROMPTS — HISTORY : ${name} ===${E}`);
    if (!versions.length) { console.log(`  ${Y}⚠  Prompt introuvable${E}`); return; }
    versions.forEach(v => {
      const cur = v.is_current ? G+' ← current'+E : '';
      console.log(`  v${v.version}  ${v.created_at?.slice(0,10)}  ${v.note?.slice(0,40)}${cur}`);
    });
  } else if (action === 'rollback') {
    if (!name) { console.log(`  Usage: observability prompts rollback <name>`); return; }
    try {
      const prev = ps.rollback(name);
      console.log(`  ${G}✓ ${name} rollback → v${prev}${E}`);
    } catch (e) { console.error(`  ${R}✗ ${e.message}${E}`); }
  } else {
    console.log(`\n  Sous-commandes prompts: list, history <name>, rollback <name>`);
  }
}

// ── REPORT ────────────────────────────────────────────────────────────────────
function cmdReport() {
  const traces  = tracer.loadTraces();
  const cbStatus = cb.getStatus();
  const prompts = ps.listAll();

  let totalCalls = traces.length, totalErrors = traces.filter(t => !t.success).length;
  let totalMs = traces.reduce((s, t) => s + (t.duration_ms||0), 0);
  const avgMs = totalCalls ? Math.round(totalMs / totalCalls) : 0;

  const html = `<!DOCTYPE html>
<html lang="fr"><head><meta charset="UTF-8"><title>Observability Report</title>
<script src="https://cdn.jsdelivr.net/npm/chart.js"></script>
<style>body{font-family:sans-serif;background:#0f172a;color:#e2e8f0;padding:24px}
h1{color:#7c3aed}h2{color:#94a3b8;font-size:1rem;margin-top:24px}
.card{background:#1e293b;border-radius:8px;padding:16px;margin:12px 0}
.grid{display:grid;grid-template-columns:repeat(auto-fit,minmax(150px,1fr));gap:12px;margin:16px 0}
.metric{background:#1e293b;border-radius:8px;padding:16px;text-align:center}
.metric .val{font-size:1.8rem;font-weight:700;color:#7c3aed}
.metric .lbl{font-size:.75rem;color:#64748b;margin-top:4px}
.ok{color:#4ade80}.warn{color:#fbbf24}.err{color:#f87171}
table{width:100%;border-collapse:collapse}td,th{padding:8px;border-bottom:1px solid #1e293b;font-size:.85rem}
th{color:#64748b;text-align:left}</style></head>
<body>
<h1>🔭 Observability Report</h1>
<p style="color:#64748b">Généré le ${new Date().toLocaleString('fr-FR')}</p>
<div class="grid">
  <div class="metric"><div class="val">${totalCalls}</div><div class="lbl">Appels LLM</div></div>
  <div class="metric"><div class="val ${totalErrors>0?'err':'ok'}">${totalErrors}</div><div class="lbl">Erreurs</div></div>
  <div class="metric"><div class="val">${avgMs}ms</div><div class="lbl">Durée moy.</div></div>
  <div class="metric"><div class="val">${cbStatus.state==='CLOSED'?'<span class="ok">CLOSED</span>':cbStatus.state==='OPEN'?'<span class="err">OPEN</span>':'<span class="warn">HALF</span>'}</div><div class="lbl">Circuit Breaker</div></div>
  <div class="metric"><div class="val">${cbStatus.cacheSize}</div><div class="lbl">Cache LLM</div></div>
</div>
<div class="card"><h2>Appels LLM récents (20 derniers)</h2>
<table><tr><th>Date</th><th>Agent</th><th>Fonction</th><th>Durée</th><th>Statut</th></tr>
${traces.slice(-20).reverse().map(t=>`<tr>
  <td>${t.ts?.slice(11,19)||''}</td><td>${t.agent}</td><td>${t.fn}</td>
  <td class="${t.duration_ms>SLOW_MS?'err':t.duration_ms>3000?'warn':'ok'}">${Math.round(t.duration_ms||0)}ms</td>
  <td class="${t.success?'ok':'err'}">${t.success?'OK':'ERR'}</td>
</tr>`).join('')}
</table></div>
${prompts.length?`<div class="card"><h2>Prompts versionnés (${prompts.length})</h2>
<table><tr><th>Nom</th><th>Version</th><th>Appels</th><th>Description</th></tr>
${prompts.map(p=>`<tr><td>${p.name}</td><td>v${p.current_version}</td><td>${p.metrics?.calls||0}</td><td style="color:#94a3b8">${p.description?.slice(0,50)||''}</td></tr>`).join('')}
</table></div>`:''}
</body></html>`;

  const outPath = path.join(DOCS_DIR, 'observability-report.html');
  fs.writeFileSync(outPath, html, 'utf8');
  console.log(`  ${G}✓ ${outPath}${E}`);
}

// ── Main ──────────────────────────────────────────────────────────────────────
function main() {
  const args = process.argv.slice(2).filter(a => !a.startsWith('--'));
  const [cmd, sub, arg3] = args;

  switch (cmd) {
    case 'metrics':   cmdMetrics();         break;
    case 'anomalies': cmdAnomalies();       break;
    case 'cost':      cmdCost();            break;
    case 'cb':        cmdCb(sub);           break;
    case 'prompts':   cmdPrompts(sub, arg3);break;
    case 'report':    cmdReport();          break;
    case 'clear':     tracer.clearTraces(); console.log(`  ${G}✓ Traces effacées${E}`); break;
    default:
      console.log(`
${B}Observability Agent${E} — Monitoring LLM + Résilience + Prompts

  metrics             Métriques des appels LLM (appels, ms, P95, erreurs)
  anomalies           Détecte appels lents (>${SLOW_MS}ms) et bursts d'erreurs
  cost                Estimation coût tokens (Groq pricing)
  cb status           État du Circuit Breaker (CLOSED/OPEN/HALF_OPEN)
  cb reset            Réinitialise le Circuit Breaker
  cb cache            Statistiques du cache LLM
  prompts list        Liste les prompts versionnés
  prompts history <n> Historique des versions d'un prompt
  prompts rollback <n>Rollback à la version précédente
  report              Génère docs/observability-report.html
  clear               Vide logs/traces.jsonl

Coût LLM : $0 — 100% déterministe
`);
  }
}

main();
